dgp/utils/render_3d_to_2d.py

```diff
@@ -0,0 +1,422 @@
+    """
+    bboxes_2d = defaultdict(lambda: defaultdict(list))
+    if len(dataset):
+        if max_num_items is not None:
+            if max_num_items > len(dataset):
+                logging.info(
+                    "`max_num_items` is reduced to the dataset size, from {:d} to {:d}".format(
+                        max_num_items, len(dataset)
+                    )
+                )
+                max_num_items = len(dataset)
+
+        ontology_table = dataset.dataset_metadata.ontology_table
+
+        if ANNOTATIONS_2D in ontology_table:
+            ontology = ontology_table[ANNOTATIONS_2D]
+
+        if camera_datum_names is None:
+            camera_datum_names = sorted(dataset.list_datum_names_available_in_all_scenes(datum_type='image'))
+
+        st = time.time()
+        logging_span = 200
+        for idx, datums in enumerate(dataset):
+            # no temporal context
+            datums = datums[0]
+            if idx == max_num_items:
+                break
+            if idx % logging_span == 0:
+                logging.info(f"2D:Frame {idx + 1} of {len(dataset)} in {time.time() - st:.2f}s.")
+            datums = {datum['datum_name']: datum for datum in datums}
+            camera_datums = [(camera_datum_name, datums[camera_datum_name]) for camera_datum_name in camera_datum_names]
+            # Visualize bounding box 2d
+            if ANNOTATIONS_2D in dataset.requested_annotations:
+                for camera_datum_name, camera_datum in camera_datums:
+                    for bbox in camera_datum[ANNOTATIONS_2D]:
+                        bboxes_2d[camera_datum_name][ontology.contiguous_id_to_name[bbox.class_id]].append(bbox.ltwh)
+    return bboxes_2d
+
+
+def associate_lidar_and_camera_2d_bboxes(
+    bboxes_from_camera: list[np.ndarray], bboxes_from_lidar: list[Tuple[np.ndarray, np.ndarray]]
+) -> list[Tuple]:
+    """Associate 3d bounding boxes and 2d bounding boxes to the same object by checking
+       whether the projected centroid of 3d bounding box is inside the 2d bounding box or not.
+    Parameters
+    ----------
+    bboxes_from_camera: list[np.ndarray]
+        A list of 2d bounding boxes.
+    bboxes_from_lidar: list[Tuple[np.ndarray, np.ndarray]]
+        A list of Tuple (centroid in 2d image, 3d bounding box).
+    Returns
+    -------
+    associated: list[Tuple[bounding_box_2d, bounding_box_3d, centroid_2d]]
+        a dictionary with key is the camera name, value is a dictionary whose key is class_name of bounding_box_2d,
+        value is list of boxes as (N, 4) np.ndarray in format ([left, top, width, height])
+    """
+    associated = []
+    if bboxes_from_camera and bboxes_from_lidar:
+        for bbox_camera in bboxes_from_camera:
+            l, t, w, h = bbox_camera
+            for bbox_lidar, bbox_centroid_2d in bboxes_from_lidar:
+                if bbox_centroid_2d is None:
+                    continue
+                bbox_centroid_x, bbox_centroid_y = bbox_centroid_2d[0]
```

Comment:
I think you could do this in one line if the inputs are numpy arrays, see https://github.com/TRI-ML/dgp/blob/master/dgp/utils/camera.py#L43